lib/instrument/transformer/nskeyed.js

```diff
@@ -0,0 +1,431 @@
+    this.objects = plist.$objects;
+  }
+
+  /**
+   * use the UNARCHIVE_CLASS_MAP to find the unarchiving delegate of a uid
+   */
+  classForUid (index) {
+    const meta = this.objects[index.UID];
+    const name = meta.$classname;
+    const klass = UNARCHIVE_CLASS_MAP[name];
+    if (!klass) {
+      throw new Error(`no mapping for ${name} in UNARCHIVE_CLASS_MAP`);
+    }
+    return klass;
+  }
+
+  decodeKey (obj, key) {
+    const val = obj[key];
+    return _.isNil(val?.UID) ? val : this.decodeObject(val);
+  }
+
+  decodeObject (index) {
+    if (index === NULL_UID) {
+      return null;
+    }
+    const obj = this.unpackedUids[index];
+    if (obj === CYCLE_TOKEN) {
+      throw new Error(`archive has a cycle with ${index}`);
+    }
+    if (!_.isUndefined(obj)) {
+      return obj;
+    }
+    const rawObj = this.objects[index.UID];
+    this.unpackedUids[index.UID] = CYCLE_TOKEN;
+
+    if (!rawObj?.$class) {
+      this.unpackedUids[index.UID] = obj;
+      return rawObj;
+    }
+    const klass = this.classForUid(rawObj.$class);
+    const klassObj = new klass().decodeArchive(new ArchivedObject(rawObj, this));
+    this.unpackedUids[index.UID] = klassObj;
+    return klassObj;
+  }
+
+  toObject () {
+    this.unpackArchiveHeader();
+    return this.decodeObject(this.topUID);
+  }
+}
+
+/**
+ * Capable of packing an object tree into the NSKeyedArchive format.
+ * Apple's implementation can be found here:
+ * https://github.com/apple/swift-corelibs-foundation/blob/main/Sources/Foundation/NSKeyedArchiver.swift
+ */
+class Archive {
+  constructor (inputObject) {
+    this.input = inputObject;
+    this.classMap = {};
+    this.objects = ['$null']; // objects that go directly into the archive, always start with $null
+  }
+
+  uidForArchiver (archiver) {
+    let val = this.classMap[archiver];
+    if (val) {
+      return new plistlib.UID(val);
+    }
+    this.classMap[archiver] = this.objects.length;
+    val = new plistlib.UID(this.objects.length);
+    this.objects.push({
+      '$classes': [archiver],
+      '$classname': archiver
+    });
+    return val;
+  }
+
+  archive (obj) {
+    if (_.isUndefined(obj) || _.isNull(obj)) {
+      return NULL_UID;
+    }
+    const index = new plistlib.UID(this.objects.length);
+    if (PRIMITIVE_TYPES.includes(obj.constructor.name)) {
+      this.objects.push(obj);
+      return index;
+    }
+    const archiveObj = {};
+    this.objects.push(archiveObj);
+    this.encodeTopLevel(obj, archiveObj);
+    return index;
+  }
+
+  encode (val) {
+    if (NON_ENCODABLE_TYPES.includes(typeof val)) {
+      return val;
+    }
+    return this.archive(val);
+  }
+
+  encodeTopLevel (obj, archiveObj) {
+    if (obj instanceof Array) {
+      return this.encodeArray(obj, archiveObj);
+    } else if (obj instanceof Set) {
+      return this.encodeSet(obj, archiveObj);
+    } else if (obj instanceof Object) {
+      const objName = obj.constructor.name;
+      if (objName in UNARCHIVE_CLASS_MAP) {
```

Comment:
can we rely on this condition? not all keys in this dictionary match to class names

Reply:
Only special class/instance are useful, such as NSURL, NSUUID, NSDate .And this class must also have the encodeArchive method

Reply:
makes sense. Perhaps it would be easier to read this could if we had the above explanation as a comment there